# Project 9 -- Jacqueline Sanchez

**TA Help:** N/A

- 
    
**Collaboration:** N/A
    
- 
- 

## Question 1

In [1]:
# reading the ice cream `products` and `reviews` files into seperate df 
library(data.table)
productsDF <- fread("/anvil/projects/tdm/data/icecream/combined/products.csv")
reviewsDF <- fread("/anvil/projects/tdm/data/icecream/combined/reviews.csv")

# Display the heads
head(productsDF, 2)
head(reviewsDF, 2)

dim(productsDF)
dim(reviewsDF)
# head

brand key  name                          
1 bj    0_bj Salted Caramel Core           
2 bj    1_bj Netflix & Chilll'd\342\204\242
  subhead                                                                   
1 Sweet Cream Ice Cream with Blonde Brownies & a Salted Caramel Core        
2 Peanut Butter Ice Cream with Sweet & Salty Pretzel Swirls & Fudge Brownies
  description                                                                                                                                                                                                                                                                                                                                         
1 Find your way to the ultimate ice cream experience with our Cores. Whether your primal urges lead you to the center of salted caramel or directly to the blonde brownies, you\342\200\231ll be in total control of your own ice cream destiny.                                                                                                      
2 There\342\200\231s something for everyone to watch on Netflix & flavors for everyone to enjoy from Ben & Jerry\342\200\231s, so we\342\200\231ve teamed up to bring you a chillaxing new creation that\342\200\231s certain to satisfy any sweet or salty snack craving. It\342\200\231s a flavorful world, and everyone is invited to grab a spoon.
  rating rating_count
1 3.7    208         
2 4.0    127         
  ingredients                                                                                                                                                                                                                                                                                                                                                                  
1 CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER), WATER, BROWN SUGAR, SUGAR, MILK, WHEAT FLOUR, EGG YOLKS, CORN SYRUP, EGGS, BUTTER (CREAM, SALT), BUTTEROIL, PECTIN, SEA SALT, SOYBEAN OIL, VANILLA EXTRACT, GUAR GUM, SOY LECITHIN, BAKING POWDER (SODIUM ACID PYROPHOSPHATE, SODIUM BICARBONATE, CORN STARCH, MONOCALCIUM PHOSPHATE), BAKING SODA, SALT, CARRAGEENAN, LACTASE
2 CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER), WATER, SUGAR, PEANUTS, WHEAT FLOUR, CANOLA OIL, EGG YOLKS, CORN STARCH, PEANUT OIL, COCOA POWDER, SALT, SOYBEAN OIL, INVERT CANE SUGAR, MILK FAT, EGGS, EGG WHITES, GUAR GUM, SOY LECITHIN, TAPIOCA STARCH, BAKING SODA, CARRAGEENAN, VANILLA EXTRACT, BARLEY MALT, MALTED BARLEY FLOUR

brand key  author         date       stars
1 bj    0_bj Ilovebennjerry 2017-04-15 3    
2 bj    0_bj Sweettooth909  2020-01-05 5    
  title                                   helpful_yes helpful_no
1 Not enough brownies!                    10          3         
2 I\342\200\231m OBSESSED with this pint!  3          0         
  text                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         
1 Super good, don't get me wrong. But I came for the caramel and brownies, not the sweet cream. The packaging made it seem like brownies were packed and bountiful *crying frowny emoji* I'd say the taste of this was amazing, but the ratio of brownie to sweet cream was disappointing. Liked it regardless but probably won't buy again simply because it didn't live up to its promising package. I'll find another one that has a better ratio and wayyy more yummy chewy brownies.\nOverall, good flavor, texture, idea, and brownies. Not so great caramel/sweet cream/ brownie RATIO. Just add more brownies. Please.                                                                                                                 
2 I decided to try it out although I\342\200\231m not a huge caramel fan, and the first buy was ok, didn\342\200\231t like the caramel too much and for some reason that specific pint barely had any brownies! Like there were some on top but from the middle to the bottom? Zilch. Nada. Nothing! It was disappointing. But for some reason, I bought it again and I really do believe it was just that point bcuz the second one was glorious!! I had a big brownie chunk in EVERY bite. Plus the caramel didn\342\200\231t bother me as much and I took a liking to it. Conclusion: I\342\200\231m on my fifth pint, of the week and I\342\200\231m going back to get my daily pint cus I\342\200\231m kinda sick and need some happiness.
  taste ingredients texture likes
1 NA    NA          NA           
2 NA    NA          NA

[1] 241   8

[1] 21674    13

In [2]:
# just analyzing and exploring ways to find similar columns between sets
# Approach 1: using `intersect` function
# specifying the columns of interest
columns_to_check <- c("brand", "key", "ingredients")

# finding shared columns
shared_columns <- intersect(columns_to_check, colnames(productsDF))
shared_columns_reviews <- intersect(columns_to_check, colnames(reviewsDF))

# displaying shared columns
shared_columns
shared_columns_reviews

[1] "brand"       "key"         "ingredients"

[1] "brand"       "key"         "ingredients"

In [3]:
# Approach 2: using pipeline operator
# check which specified columns are present in each data frame
products_shared <- columns_to_check %in% colnames(productsDF)
reviews_shared <- columns_to_check %in% colnames(reviewsDF)

# display results
data.frame(Column = columns_to_check, 
           In_Products = products_shared, 
           In_Reviews = reviews_shared)

Column      In_Products In_Reviews
1 brand       TRUE        TRUE      
2 key         TRUE        TRUE      
3 ingredients TRUE        TRUE

In [4]:
# merging both the df according to the `brand` and key values 
newmergedDF <- merge(productsDF, reviewsDF, by = c("brand", "key"))
dim(newmergedDF)
# colnames(newmergedDF)
names(newmergedDF)

[1] 21674    19

[1] "brand"         "key"           "name"          "subhead"      
 [5] "description"   "rating"        "rating_count"  "ingredients.x"
 [9] "author"        "date"          "stars"         "title"        
[13] "helpful_yes"   "helpful_no"    "text"          "taste"        
[17] "ingredients.y" "texture"       "likes"

In [5]:
head(newmergedDF, 1)

brand key  name               
1 bj    0_bj Salted Caramel Core
  subhead                                                           
1 Sweet Cream Ice Cream with Blonde Brownies & a Salted Caramel Core
  description                                                                                                                                                                                                                                   
1 Find your way to the ultimate ice cream experience with our Cores. Whether your primal urges lead you to the center of salted caramel or directly to the blonde brownies, you\342\200\231ll be in total control of your own ice cream destiny.
  rating rating_count
1 3.7    208         
  ingredients.x                                                                                                                                                                                                                                                                                                                                                                
1 CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER), WATER, BROWN SUGAR, SUGAR, MILK, WHEAT FLOUR, EGG YOLKS, CORN SYRUP, EGGS, BUTTER (CREAM, SALT), BUTTEROIL, PECTIN, SEA SALT, SOYBEAN OIL, VANILLA EXTRACT, GUAR GUM, SOY LECITHIN, BAKING POWDER (SODIUM ACID PYROPHOSPHATE, SODIUM BICARBONATE, CORN STARCH, MONOCALCIUM PHOSPHATE), BAKING SODA, SALT, CARRAGEENAN, LACTASE
  author         date       stars title                helpful_yes helpful_no
1 Ilovebennjerry 2017-04-15 3     Not enough brownies! 10          3         
  text                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
1 Super good, don't get me wrong. But I came for the caramel and brownies, not the sweet cream. The packaging made it seem like brownies were packed and bountiful *crying frowny emoji* I'd say the taste of this was amazing, but the ratio of brownie to sweet cream was disappointing. Liked it regardless but probably won't buy again simply because it didn't live up to its promising package. I'll find another one that has a better ratio and wayyy more yummy chewy brownies.\nOverall, good flavor, texture, idea, and brownies. Not so great caramel/sweet cream/ brownie RATIO. Just add more brownies. Please.
  taste ingredients.y texture likes
1 NA    NA            NA

Interesting! like mentioned in the `The Examples Book` 
the dimenstion for productsDF is 241*8 and reviewsDF is 21674 * 13. 
How come when we merged them we get only 19 columns? Shouldn't it be 21?<br>
brand and key are the same except for ingredient since in both df they play different roles these were not merged. 

In [6]:
# new df is a subset, where there are 978 rows that correspond to the name "Chocolate Chip Chip Dough" , brand "bj" and key "16_bj"
ChocolateChipCookieDoughDF <- subset(newmergedDF, (brand == "bj") & (key == "16_bj"))
head(ChocolateChipCookieDoughDF,1)

brand key   name                       
1 bj    16_bj Chocolate Chip Cookie Dough
  subhead                                                   
1 Vanilla Ice Cream with Gobs of Chocolate Chip Cookie Dough
  description                                                                                                                                                                                                    
1 We knew we were onto something big when we made the world\342\200\231s first batch of Chocolate Chip Cookie Dough ice cream in 1984. Today the flavor still reigns among our all-time most popular concoctions.
  rating rating_count
1 4.6    983         
  ingredients.x                                                                                                                                                                                                                                                                                                                                       
1 CREAM, SKIM MILK, LIQUID SUGAR (SUGAR, WATER), WATER, WHEAT FLOUR, SUGAR, BROWN SUGAR, EGG YOLKS, BUTTER (CREAM, SALT), EXPELLER PRESSED SOYBEAN OIL, EGGS, COCONUT OIL, CHOCOLATE LIQUOR, VANILLA EXTRACT, COCOA (PROCESSED WITH ALKALI), COCOA POWDER, SALT, MOLASSES, GUAR GUM, COCOA BUTTER, NATURAL FLAVOR, CARRAGEENAN, MILK FAT, SOY LECITHIN
  author   date       stars title                                      
1 Justamom 2018-05-08 2     Hey Ben and Jerry where is my cookie dough?
  helpful_yes helpful_no
1 5           1         
  text                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                
1 Always have loved your chocolate chip cookie dough ice cream, my husband knows this and came home with some. Well it takes very little to excite me these days but after a day of errands, housework, homework, dinner then getting son to bed all I kept thinking of, was that small tub of heaven in my freezer. So finally I put on some Guilty pleasure reality television snugged into my couch with my favorite blanket and my favorite ice cream only to notice that after a couple spoonfuls......NO COOKIE DOUGH! This can\342\200\231t be correct NOOOOOOOOO!! First I was really doubting I was smarter than a fifth grader and now NO COOKIE DOUGH! Not the ending I needed to this day!
  taste ingredients.y texture likes
1 NA    NA            NA

In [7]:
names(ChocolateChipCookieDoughDF)
dim(ChocolateChipCookieDoughDF)

[1] "brand"         "key"           "name"          "subhead"      
 [5] "description"   "rating"        "rating_count"  "ingredients.x"
 [9] "author"        "date"          "stars"         "title"        
[13] "helpful_yes"   "helpful_no"    "text"          "taste"        
[17] "ingredients.y" "texture"       "likes"

[1] 978  19

In [8]:
# printing the average number of stars for the 978 reviews in the df
mean(ChocolateChipCookieDoughDF$rating)

[1] 4.6

* The average number of starts for the 978 review in the df is a rating of **4.6 stars**. So, they have really good over all ratings

## Question 2

In [9]:
# code here
# find averagae num of `stars` for the reiviews of ice cream with  `name` == "Half Baked\302\256"
specific_reviews <- newmergedDF[newmergedDF$name == "Half Baked\302\256", ]

# checking filtered reviews 
# print(specific_reviews)

# Calculate the average number of stars
average_stars <- mean(specific_reviews$stars, na.rm = TRUE)
print(average_stars)

[1] 4.695701


The average number of stars for the review of ice cream with the name `Half Baked\302\256` is around **4.6957 stars**

## Question 3

In [10]:
# code here
# Find the products that have "Chocolate" in the name
chocolate_products <- newmergedDF[grepl("Chocolate", newmergedDF$name, ignore.case = TRUE), ]

# checking the filtered data
# print(chocolate_products)

# average number of stars for items that contain chocolate 
average_stars_chocolate <- mean(chocolate_products$stars, na.rm = TRUE)

# average stars
print(average_stars_chocolate)

[1] 4.154539


Seems like chocolate products are really popular with a good score of **4.154 stars**

## Question 4

In [11]:
# code here
# reading two separate data frames, it is said thta they do not have any columns in common 
library(data.table)
airportsDF <- read.csv("/anvil/projects/tdm/data/flights/subset/airports.csv", stringsAsFactors = FALSE)
flightdataDF <- fread("/anvil/projects/tdm/data/flights/subset/1990.csv")

In [12]:
# but these two df have some features that might match the other
# in airportDF has an attribute "iata" which contain three-letter codes corresponding to airports
# similar to `origin` and `Dest` attributes found in flightdataDF
# so we are merging iata and Origin which studies when the flight departs
mybigDF <- merge(airportsDF, flightdataDF, by.x = "iata", by.y = "Origin")
head(mybigDF, 2)

iata airport                     city      state country lat      long    
1 ABE  Lehigh Valley International Allentown PA    USA     40.65236 -75.4404
2 ABE  Lehigh Valley International Allentown PA    USA     40.65236 -75.4404
  Year Month DayofMonth ... TaxiIn TaxiOut Cancelled CancellationCode Diverted
1 1990 9     26         ... NA     NA      0         NA               0       
2 1990 9     28         ... NA     NA      0         NA               0       
  CarrierDelay WeatherDelay NASDelay SecurityDelay LateAircraftDelay
1 NA           NA           NA       NA            NA               
2 NA           NA           NA       NA            NA

In [13]:
# filtering only flights made in Indiana
Indiana_flights <- mybigDF[mybigDF$state == "IN", ]
# str(Indiana_flights)
# names(Indiana_flights)

# calculating the average departure delay
average_dep_delay <- mean(Indiana_flights$DepDelay, na.rm = TRUE)

# avg dep delay
print(average_dep_delay)

[1] 5.370691


The average departure delay for all flights whose origin is **Indiana** is aroud **5.3707**

## Question 5

In [14]:
# filtering only flights made in Texas
Texas_flights <- mybigDF[mybigDF$state == "TX", ]
# names(Texas_flights)
# calculating the average departure delay
average_dep_delay <- mean(Texas_flights$DepDelay, na.rm = TRUE)
# avg dep delay
print(average_dep_delay)

[1] 7.802326


The average departure delay for all flights whose origin is **Texas** is aroud **7.8023**

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.